src/filters/oidc/oidc_filter.cc

```diff
@@ -105,6 +108,44 @@ google::rpc::Code OidcFilter::Process(
   // then let request continue.
   // (It is up to the downstream system to validate the header is valid.)
   if (headers.contains(idp_config_.id_token().header())) {
+    const auto &header_value = headers[idp_config_.id_token().header()];
+
+    if (!absl::StartsWith(header_value, idp_config_.id_token().preamble())) {
+      spdlog::info(
+          "{}: ID token header value doesn't match with the configured "
+          "preamble value",
+          __func__);
+      return google::rpc::Code::UNAUTHENTICATED;
+    }
+
+    if (!session_id_optional.has_value()) {
```

Comment:
> We believe that the benefit of not validating the JWT nonce value

Not sure I follow. I think the points @bburky is making is not against validating the nonce/state values. Instead we are proposing the validation of the id token from header is not necessary. This I agree. authservice itself today already manages the extracting the session info and id token in its session store. We haven't documented any programmatically access via passing id token in request directly.

I think this can be done/discussed separately and having this PR moving forwarded AFAIK, so creating another issue to continue and simplify. https://github.com/istio-ecosystem/authservice/issues/189

Let me know if you have different thoughts, thanks.

Reply:
> I honestly think this functionality could be completely removed. Just drop any incoming ID Token headers

+1. This simplifies the things a lot.

@Shikugawa I think Blake's point is that we don't have to consider id token from headers at all. We always expect the flow is kicked from end user using a browswer. Therefore considering session id itself is enough. We can remove the check of the id token from header.

To be more precise, before the PR, the flow is as following:

1. Check if there's already id token from Authorization header.
2. If so, allow the request to go through, __even not checking the id token's validity__. --> This is the problem we see.
3. If no token, we extracts the session id from the cookie and handle request from that point on.

We don't need step 1 & 2, so the intended flow should be:

1. Get the session id from the requests.
2. Extract the current session store, and populates the response header "Authorization: Bearer"
3. Check the id token with nonce, id token etc.

Reply:
fixed